File: javascript/778-Swim-in-Rising-Water.js
/**
 * https://leetcode.com/problems/swim-in-rising-water/
 * @param {number[][]} grid
 * @return {number}
 */
 var swimInWater = (grid) => {
    const seen = new Set();
    const minHeap = getHeap(grid);

    return getTime(grid, seen, minHeap);
};


const getHeap = (grid, minHeap = new MinPriorityQueue()) => {
    minHeap.enqueue([ 0, 0 ], grid[0][0]);

    return  minHeap;
}

var getTime = (grid, seen, minHeap, maxTime = 0) => {
    const [ rows, cols ] = [ (grid.length - 1), (grid[0].length - 1) ];

    while (!minHeap.isEmpty()) {
        const { element, priority: cost } = minHeap.dequeue();
        const [ row, col ] = element;

        seen.add(grid[row][col]);
        maxTime = Math.max(maxTime, cost);

        const isEnd = (row === rows) && (col === cols);
        if (isEnd) return maxTime;

        checkNeighbors(grid, row, rows, col, cols, seen, minHeap);
    }
}

var checkNeighbors = (grid, row, rows, col, cols, seen, minHeap) => {
    for (const [ _row, _col ] of getNeighbors(row, rows, col, cols)) {
        if (seen.has(grid[_row][_col])) continue;

        minHeap.enqueue([ _row, _col ], grid[_row][_col]);
    }
}

const getNeighbors = (row, rows, col, cols) => [ [1, 0], [-1, 0], [0, 1], [0, -1] ]
    .map(([ _row, _col ]) => [ (row + _row), (col + _col) ])
    .filter(([ _row, _col ]) => ((0 <= _row) && (_row <= rows) && (0 <= _col) && (_col <= cols)))

File: java/778-Swim-in-Rising-Water.java
// Solution: Greedy Approach with Min Heap
// Time Complexity: O((n^2)*log(n))
class Solution {

    private int[][] dirs = { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 } };

    public int swimInWater(int[][] grid) {
        int len = grid.length;

        if (len == 1) {
            return 0;
        }

        var seen = new boolean[len][len];
        seen[0][0] = true;

        var minHeap = new PriorityQueue<Integer[]>((a, b) -> a[0] - b[0]);
        minHeap.add(new Integer[] { grid[0][0], 0, 0 });

        int result = 0;

        while (!minHeap.isEmpty()) {
            var curr = minHeap.poll();

            result = Math.max(result, curr[0]);

            if (curr[1] == len - 1 && curr[2] == len - 1) {
                break;
            }

            for (int i = 0; i < 4; i++) {
                int x = curr[1] + dirs[i][0];
                int y = curr[2] + dirs[i][1];

                if (x < 0 || x >= len || y < 0 || y >= len || seen[x][y]) {
                    continue;
                }

                minHeap.add(new Integer[] { grid[x][y], x, y });
                seen[x][y] = true;
            }
        }

        return result;
    }
}
